File: Clustering/DBSCAN.cpp
#include <cmath>
#include <functional>
#include <iostream>
#include <map>
#include <stack>
#include <unordered_map>
#include <vector>

enum PointType : int {
    kpointType_UNDO = 0,
    kpointType_CORE = 3,
    kpointType_BORDER = 2,
    kpointType_NOISE = 1
};

enum class ParticleType : int {
    kparticleType_UNKNOWN = 0,         // 未分类
    kparticleType_StopLine = 1,        // 停止线
    kparticleType_CrossWalk = 2,       // 人行横道
    kparticleType_LaneLine = 3,        // 车道线
    kparticleType_RoadEdge = 4,        // 路沿
    kparticleType_Junction = 5,        // bev_junction
    kparticleType_TrafficLight = 6,    // 交通灯
    kparticleType_CarStreamPoint = 7,  // 专线点
    kparticleType_OutLine = 8,         // 轮廓线
};

enum class keyVal : int {
    keyVal_THETA = 0,       // 向量夹角，正反向会区分，范围0-pi
    keyVal_THETA_HALF = 5,  // 直线夹角，正反向不区分，范围0-pi/2
    keyVal_X = 1,           // x坐标
    keyVal_Y = 2,           // y坐标
    keyVal_CENTER = 3,      // 欧式距离
    keyVal_ALL = 4,
    keyVal_LATERAL = 6,        // 计算两条线之间的横向距离
    keyVal_CLUSTER = 7,        // 计算两个cluster之间的距离，依据line之间的距离
    keyVal_DISCRETECOMPS = 8,  // 计算两个离散要素(只支持人行横道和停止线)的距离
    keyVal_MIN_X_Y = 9,
    keyVal_RADIUS = 10,
};

struct Vec2D {
    float x;
    float y;
    Vec2D() = default;
    explicit Vec2D(float x, float y) : x(x), y(y) {}
};

struct Particle {
    std::string obj_id;
    float x = 0.0f;
    float y = 0.0f;
    float weight = 0.0f;
    std::vector<float> xn;                                             // 多维数据
    int cluster = 0;                                                   // 第几个簇
    PointType point_type = PointType::kpointType_UNDO;                 // 1 noise 2 border 3 core
    ParticleType particle_type = ParticleType::kparticleType_UNKNOWN;  // particle语义类型
    int pts = 0;                                                       // points in MinPts
    int corePointID = -1;                                              // 核心点的标号
    std::vector<int> corepts;
    int visited = 0;
    int cluster_idx = 0;  // 用来聚类logical cluster时记录idx
    // spTrafficLightInstance tfl_instance;
    Particle() = default;

    explicit Particle(std::string id, float x, float y, float weight, ParticleType type)
        : obj_id(id), x(x), y(y), weight(weight), particle_type(type) {}
    void CalculateWeight(const Vec2D &scan_start_points, float scan_radius);
    void SetParticleType(const int &type);
};

float squareDistance(const Particle &a, const Particle &b) {
    return std::sqrt((a.x - b.x) * (a.x - b.x) + (a.y - b.y) * (a.y - b.y));
}

void DBSCAN(std::vector<Particle> &dataset, keyVal keyValType, const float &Eps,
            const int &MinPts) {
    int clusterID = 0;
    size_t len = dataset.size();
    if (len == 0U) {
        return;
    }
    static std::map<keyVal, std::function<float(const Particle &, const Particle &)>> distFuncMap =
        {{keyVal::keyVal_CENTER, squareDistance}};
    auto iter = distFuncMap.find(keyValType);
    std::vector<std::vector<float>> distP2P(len, std::vector<float>(len, 0.0f));
    std::vector<Particle> corePoint;
    if (iter != distFuncMap.end()) {
        for (int i = 0; static_cast<size_t>(i) < len; i++) {
            // pts添加自己
            dataset[i].pts++;
            for (int j = i + 1; static_cast<size_t>(j) < len; j++) {
                float distance = (iter->second)(dataset[i], dataset[j]);
                // distP2P是对称矩阵
                distP2P[i][j] = distance;
                distP2P[j][i] = distance;
                if (distance <= Eps) {
                    dataset[i].pts++;
                    dataset[j].pts++;
                }
            }
            // core Particle 核心点，pts大于minPts的时候，该点为核心点
            if (dataset[i].pts >= MinPts) {
                dataset[i].point_type = PointType::kpointType_CORE;
                dataset[i].corePointID = i;
                corePoint.push_back(dataset[i]);
            }
        }
    } else {
        std::cout << "Invalid keyValType! " << std::endl;
    }
    // 打印出核心点
    size_t numCorePoint = corePoint.size();
    std::cout << "corePint ";
    for (auto &core : corePoint) {
        std::cout << core.obj_id << " ";
    }
    std::cout << std::endl;
    // 打印完毕

    for (size_t i = 0U; i < numCorePoint; i++) {
        std::vector<float> &dist_i = distP2P[corePoint[i].corePointID];
        std::vector<int> &corepts_i = corePoint[i].corepts;
        for (int j = 0; static_cast<size_t>(j) < numCorePoint; j++) {
            float distTemp = dist_i[corePoint[j].corePointID];
            std::cout << "CorePoint " << corePoint[i].obj_id << " corepts: ";
            if (distTemp <= Eps) {
                // other point orderID link to core point
                std::cout << corePoint[j].obj_id << " ";
                corepts_i.push_back(j);
            }
            std::cout << std::endl;
            // 把每一个在核心点领域的核心点放到一起
        }
    }
    for (size_t i = 0U; i < numCorePoint; i++) {
        // 遍历所有的核心点
        std::stack<Particle *> ps;
        if (corePoint[i].visited == 1) {
            continue;
        }
        clusterID++;
        corePoint[i].cluster = clusterID;
        ps.push(&corePoint[i]);
        Particle *v;
        while (!ps.empty()) {
            v = ps.top();
            v->visited = 1;
            ps.pop();
            for (size_t j = 0U; j < v->corepts.size(); j++) {
                // 最开始归类的一簇进行遍历
                if (corePoint[v->corepts[j]].visited == 1) {
                    continue;
                }
                corePoint[v->corepts[j]].cluster = corePoint[i].cluster;
                corePoint[v->corepts[j]].visited = 1;
                ps.push(&corePoint[v->corepts[j]]);
            }
        }
    }

    // border point,joint border point to core point
    // k用来在dataset中统计是第几个核心点
    int k = 0;
    for (size_t i = 0U; i < len; i++) {
        if (dataset[i].point_type == PointType::kpointType_CORE) {
            // 如果该点是核心点，在上面已经访问过了，就不再访问，
            // 因为核心点不可能是边界点，没必要再访问一次
            dataset[i].cluster = corePoint[k++].cluster;
            // 遍历到第k个核心点时，把属于的簇id给原来的dataset
            continue;
        }
        for (size_t j = 0U; j < numCorePoint; j++) {
            float distTemp = distP2P[i][corePoint[j].corePointID];
            if (distTemp <= Eps) {
                dataset[i].point_type = PointType::kpointType_BORDER;
                dataset[i].cluster = corePoint[j].cluster;
                break;
            }
        }
    }
}

void GetCluster(std::vector<Particle> &data, std::vector<std::vector<Particle>> &clusters) {
    std::unordered_map<int, std::vector<int>> cluster_ids;
    for (int i = 0; static_cast<size_t>(i) < data.size(); ++i) {
        cluster_ids[data[i].cluster].push_back(i);
    }

    for (auto &iter : cluster_ids) {
        std::vector<int> index = iter.second;
        std::vector<Particle> cluster;
        for (size_t i = 0U; i < index.size(); ++i) {
            cluster.push_back(data[index[i]]);
        }
        clusters.push_back(cluster);
    }
    return;
}

int main() {
    std::vector<Particle> dataset;
    std::vector<std::vector<Particle>> clusters;
    Particle p1("1", 1.0f, 1.0f, 1.0f, ParticleType::kparticleType_CarStreamPoint);
    Particle p2("2", 1.1f, 1.2f, 2.0f, ParticleType::kparticleType_CarStreamPoint);
    Particle p3("3", 1.5f, 1.3f, 3.0f, ParticleType::kparticleType_CarStreamPoint);
    Particle p4("4", 4.0f, 4.0f, 4.0f, ParticleType::kparticleType_CarStreamPoint);
    Particle p5("5", 5.0f, 5.0f, 5.0f, ParticleType::kparticleType_CarStreamPoint);
    Particle p6("6", 6.0f, 6.0f, 6.0f, ParticleType::kparticleType_CarStreamPoint);
    Particle p7("7", 7.0f, 7.0f, 7.0f, ParticleType::kparticleType_CarStreamPoint);
    Particle p8("8", 8.0f, 8.0f, 8.0f, ParticleType::kparticleType_CarStreamPoint);
    Particle p9("9", 9.0f, 9.0f, 9.0f, ParticleType::kparticleType_CarStreamPoint);
    Particle p10("10", 10.0f, 10.0f, 10.0f, ParticleType::kparticleType_CarStreamPoint);
    dataset.push_back(p1);
    dataset.push_back(p2);
    dataset.push_back(p3);
    dataset.push_back(p4);
    dataset.push_back(p5);
    dataset.push_back(p6);
    dataset.push_back(p7);
    dataset.push_back(p8);
    dataset.push_back(p9);
    dataset.push_back(p10);
    std::cout << "dataset size: " << dataset.size() << std::endl;
    DBSCAN(dataset, keyVal::keyVal_CENTER, 0.5f, 2);
    GetCluster(dataset, clusters);
    std::cout << "clusters size: " << clusters.size() << std::endl;
    for (size_t i = 0U; i < clusters.size(); i++) {
        std::cout << "cluster " << i << " size: " << clusters[i].size() << std::endl;
        for (size_t j = 0U; j < clusters[i].size(); j++) {
            std::cout << clusters[i][j].obj_id << " ";
        }
        std::cout << std::endl;
    }
    return 0;
}

File: Clustering/DBSCAN.py
import numpy as np
import matplotlib.pyplot as plt
import enum
import sys
from typing import Dict, Callable

# ---------------- 定义枚举和类 ----------------


class PointType(enum.Enum):
    kpointType_UNDO = 0
    kpointType_CORE = 3
    kpointType_BORDER = 2
    kpointType_NOISE = 1


class ParticleType(enum.Enum):
    kparticleType_UNKNOWN = 0
    kparticleType_LaneLine = 3


class keyVal(enum.Enum):
    keyVal_Y = 2      # y距离
    keyVal_CENTER = 3  # 欧氏距离


class Particle:
    def __init__(self, id, x, y, weight=1.0, type=ParticleType.kparticleType_UNKNOWN):
        self.obj_id = id
        self.x = x
        self.y = y
        self.weight = weight
        self.particle_type = type
        self.cluster = 0
        self.point_type = PointType.kpointType_UNDO
        self.pts = 0
        self.corePointID = -1
        self.corepts = []   # 保存 dataset 的索引
        self.visited = 0
        self.cluster_idx = 0


def squareDistance(a, b) -> float:
    return (a.x - b.x)**2 + (a.y - b.y)**2


def CalcuYDistance(a, b) -> float:
    threshold = 50.0
    if abs(a.x - b.x) > threshold:
        return sys.float_info.max
    return abs(a.y - b.y)

# ---------------- DBSCAN 算法 ----------------


def DBSCAN(dataset, KeyValType, eps, minPts):
    clusterID = 0
    length = len(dataset)
    # 没数据直接返回
    if length == 0:
        return
    # 选距离函数
    dist_func_map: Dict[keyVal, Callable[[Particle, Particle], float]] = {
        keyVal.keyVal_CENTER: squareDistance,
        keyVal.keyVal_Y: CalcuYDistance
    }
    dist_func = dist_func_map[KeyValType]

    dist_p2p = np.zeros((length, length), dtype=np.float32)
    core_point = []  # 存索引

    # Step1: 找核心点
    for i in range(length):
        dataset[i].pts = 1
        for j in range(i+1, length):
            distance = dist_func(dataset[i], dataset[j])
            dist_p2p[i][j] = distance
            dist_p2p[j][i] = distance
            if distance <= eps:
                # 小于阈值，认为此点的邻居点+1
                dataset[i].pts += 1
                dataset[j].pts += 1
        # 邻居点的个数大于阈值，认为此点为核心点
        if dataset[i].pts >= minPts:
            dataset[i].point_type = PointType.kpointType_CORE
            dataset[i].corePointID = i
            core_point.append(i)

    # Step2: 建立核心点邻居关系
    # 遍历核心点
    for i in range(len(core_point)):
        dist_i = dist_p2p[core_point[i]]
        corepts_i = dataset[core_point[i]].corepts
        for j in range(len(core_point)):
            distTemp = dist_i[core_point[j]]
            # 两个核心点的距离小于阈值
            # 将一个核心点放在另一个核心点的corepts中
            if distTemp <= eps:
                corepts_i.append(core_point[j])  # 存 dataset 索引

    # Step3: 扩展簇
    # 遍历核心点
    for i in range(len(core_point)):
        ps = []
        # 跳过已经访问过的核心点
        if dataset[core_point[i]].visited == 1:
            continue
        # 给没访问过的核心点一个簇ID
        clusterID += 1
        dataset[core_point[i]].cluster = clusterID
        ps.append(dataset[core_point[i]])
        while ps:
            # 标记此核心点为访问过
            v = ps.pop()
            v.visited = 1
            # 遍历此核心点的邻居核心点
            for idx in v.corepts:
                # 跳过已经访问过的核心点
                if dataset[idx].visited == 1:
                    continue
                # 赋予同样的簇ID，标记为访问过
                dataset[idx].cluster = clusterID
                dataset[idx].visited = 1
                ps.append(dataset[idx])

    # Step4: 标记边界点
    # 遍历所有点
    for i in range(length):
        # 如果这个点是核心点，跳过，因为核心点在第三步已经处理过了
        if dataset[i].point_type == PointType.kpointType_CORE:
            continue
        for j in range(len(core_point)):
            # distTemp为此点与第j个核心点的距离
            distTemp = dist_p2p[i][core_point[j]]
            # 如果小于阈值，认为此点为边界点（也就是这个点落在某个核心点的邻域内， 但是他本身没有足够多的邻域点）
            if distTemp <= eps:
                dataset[i].point_type = PointType.kpointType_BORDER
                dataset[i].cluster = dataset[core_point[j]].cluster
                break


# ---------------- 测试部分 ----------------
if __name__ == "__main__":
    np.random.seed(0)

    # 构造三簇随机点
    data1 = np.random.randn(30, 2) * 0.8 + np.array([-1, 4])
    data2 = np.random.randn(30, 2) * 0.3 + np.array([1, 1])
    data3 = np.random.randn(30, 2) * 0.3 + np.array([2, 3])
    data = np.vstack((data1, data2, data3))

    # 转换为 Particle
    dataset = [Particle(str(i), x, y) for i, (x, y) in enumerate(data)]

    # 运行 DBSCAN
    DBSCAN(dataset, keyVal.keyVal_CENTER, eps=0.50, minPts=4)

    # 画图
    plt.figure(figsize=(6, 6))

    for cluster_id in set(p.cluster for p in dataset):
        # 取该簇所有点
        cluster_points = [p for p in dataset if p.cluster == cluster_id]

        # 分开核心点和边界点
        core_x = [p.x for p in cluster_points if p.point_type ==
                  PointType.kpointType_CORE]
        core_y = [p.y for p in cluster_points if p.point_type ==
                  PointType.kpointType_CORE]
        border_x = [p.x for p in cluster_points if p.point_type ==
                    PointType.kpointType_BORDER]
        border_y = [p.y for p in cluster_points if p.point_type ==
                    PointType.kpointType_BORDER]

        # 用 cluster_id 控制颜色
        color = plt.cm.tab10(cluster_id % 10)

        # 核心点：大圆点
        plt.scatter(core_x, core_y, c=[
                    color], marker="o", s=80, label=f"Cluster {cluster_id} (core)")
        # 边界点：小圆点
        plt.scatter(border_x, border_y, c=[
                    color], marker=".", s=40, label=f"Cluster {cluster_id} (border)")

    # 画噪声点
    noise_x = [p.x for p in dataset if p.point_type ==
               PointType.kpointType_NOISE]
    noise_y = [p.y for p in dataset if p.point_type ==
               PointType.kpointType_NOISE]
    if noise_x:
        plt.scatter(noise_x, noise_y, c="black",
                    marker="x", s=60, label="Noise")

    plt.title("DBSCAN Clustering Result")
    plt.xlabel("X")
    plt.ylabel("Y")
    plt.legend()
    plt.show()
